### PR TITLE
Ifpack2: add option for RCM reordering in streamed KK SPILUK and KK SPTRSV in RILUK

### PR DESCRIPTION
**Pull request by vqd8a**
@trilinos/ifpack2 

## Motivation

This PR is to add a code path that allows applying RCM reordering to sub-matrices inputted to stream interfaces of KK SPILUK and KK SPTRSV in Ifpack2 RILUK. 

Because this relies on a Kokkos Kernels utility function to return RCM reordered diagonal blocks and their respective permutation indices, this PR also includes the corresponding changes in Kokkos Kernels that match the Kokkos Kernels PR https://github.com/kokkos/kokkos-kernels/pull/2125.

### CODE REVIEW

**Comment by vqd8a** [comment]
@jhux2 , @srajama1 all tests pass. Could you please review it?

File: packages/ifpack2/src/Ifpack2_RILUK_def.hpp [hunk 1]
```diff
@@ -1198,77 +1210,187 @@ apply (const Tpetra::MultiVector<scalar_type,local_ordinal_type,global_ordinal_t
...
+
+        if (mode == Teuchos::NO_TRANS) { // Solve L (D (U Y)) = X for Y.
 #if defined(KOKKOSKERNELS_ENABLE_TPL_CUSPARSE) && defined(KOKKOS_ENABLE_CUDA) && (CUDA_VERSION >= 11030)
-        //NOTE (Nov-15-2022):
-        //This is a workaround for Cuda >= 11.3 (using cusparseSpSV)
-        //since cusparseSpSV_solve() does not support in-place computation
-        MV Y_tmp (Y.getMap (), Y.getNumVectors ());
+          //NOTE (Nov-15-2022):
+          //This is a workaround for Cuda >= 11.3 (using cusparseSpSV)
+          //since cusparseSpSV_solve() does not support in-place computation
```
Comment (brian-kelley) [inline packages/ifpack2/src/Ifpack2_RILUK_def.hpp:1245]: I think this comment doesn't apply to the reordered case and should be removed

File: packages/ifpack2/src/Ifpack2_RILUK_def.hpp [hunk 2]
```diff
@@ -1198,77 +1210,187 @@ apply (const Tpetra::MultiVector<scalar_type,local_ordinal_type,global_ordinal_t
...
+            Y.elementWiseMultiply (one, *D_, Y, zero);
+          }
 
-        U_solver_->apply (Y_tmp, Y, mode); // Solve U Y = Y_tmp.
+          U_solver_->apply (ReorderedY, ReorderedY, mode); // Solve U Y = Y.
+#endif
+        }
+        else { // Solve U^P (D^P (L^P Y)) = X for Y (where P is * or T).          
+#if defined(KOKKOSKERNELS_ENABLE_TPL_CUSPARSE) && defined(KOKKOS_ENABLE_CUDA) && (CUDA_VERSION >= 11030)
+          //NOTE (Nov-15-2022):
```
Comment (brian-kelley) [inline packages/ifpack2/src/Ifpack2_RILUK_def.hpp:1273]: Same as above

File: packages/ifpack2/src/Ifpack2_RILUK_def.hpp [hunk 3]
```diff
@@ -1198,77 +1210,187 @@ apply (const Tpetra::MultiVector<scalar_type,local_ordinal_type,global_ordinal_t
...
       }
-      else { // Solve U^P (D^P (L^P Y)) = X for Y (where P is * or T).          
+      else {
+        if (mode == Teuchos::NO_TRANS) { // Solve L (D (U Y)) = X for Y.
 #if defined(KOKKOSKERNELS_ENABLE_TPL_CUSPARSE) && defined(KOKKOS_ENABLE_CUDA) && (CUDA_VERSION >= 11030)
-        //NOTE (Nov-15-2022):
-        //This is a workaround for Cuda >= 11.3 (using cusparseSpSV)
-        //since cusparseSpSV_solve() does not support in-place computation
-        MV Y_tmp (Y.getMap (), Y.getNumVectors ());
+          //NOTE (Nov-15-2022):
```
Comment (brian-kelley) [inline packages/ifpack2/src/Ifpack2_RILUK_def.hpp:1330]: Same as above

File: packages/ifpack2/src/Ifpack2_RILUK_def.hpp [hunk 4]
```diff
@@ -1198,77 +1210,187 @@ apply (const Tpetra::MultiVector<scalar_type,local_ordinal_type,global_ordinal_t
...
+            Y.elementWiseMultiply (one, *D_, Y, zero);
+          }
 
-        L_solver_->apply (Y_tmp, Y, mode); // Solve L^P Y = Y_tmp.
+          U_solver_->apply (Y, Y, mode); // Solve U Y = Y.
+#endif
+        }
+        else { // Solve U^P (D^P (L^P Y)) = X for Y (where P is * or T).          
+#if defined(KOKKOSKERNELS_ENABLE_TPL_CUSPARSE) && defined(KOKKOS_ENABLE_CUDA) && (CUDA_VERSION >= 11030)
+          //NOTE (Nov-15-2022):
```
Comment (brian-kelley) [inline packages/ifpack2/src/Ifpack2_RILUK_def.hpp:1360]: Same

**Review by brian-kelley — CHANGES_REQUESTED** [review]
There are some copy-pasted comments in the RILUK that only apply to the non-reordered case, so I would prefer that they're taken out

**Comment by vqd8a** [comment]
@brian-kelley Thanks for catching it. I removed the comments and simplified the code to have a cleaner code path for the reordered case.

**Review by brian-kelley — APPROVED** [review]

**Review by brian-kelley — APPROVED** [review]

**Review by jhux2 — APPROVED** [review]
Lgtm.  Thanks, @vqd8a.

**Comment by vqd8a** [comment]
@jhux2 Thanks for reviewing it.

**Pull request merged**